File: cesta/AMarotoChivite-cesta/AMarotoChivite-cesta-Java/src/main/java/simulacionCesta/factories/CestListFactory.java
package simulacionCesta.factories;

import simulacionCesta.models.CestList;
import simulacionCesta.models.Product;

public class CestListFactory {
    public static CestList create() {
        // Por defecto presenta un tamaño 10 de máximo
        int randomSize = 10;
        Product[] productsStorage = new Product[randomSize];

        do {
            // Introducimos productos aleatorios en la línea de cesta
            for (int i = 0; i < productsStorage.length; i++) {
                productsStorage[i] = ProductFactory.create();
            }
            // No saldremos hasta que genere por cada línea de producto un producto distinto
        } while (!checkProductNoRepeat(productsStorage));

        return new CestList(randomSize, productsStorage);
    }

    /**
     * Verificamos que se generen por cada línea de producto un producto distinto
     *
     * @param productsStorage el almacén que ha generado productos para verificar
     * @return true si no hay productos iguales, false en caso contrario
     */
    public static boolean checkProductNoRepeat(Product[] productsStorage) {
        for (int i = 0; i < productsStorage.length; i++) {
            for (int j = i + 1; j < productsStorage.length; j++) {
                if (productsStorage[i].name.equals(productsStorage[j].name)) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: cesta/AMarotoChivite-cesta/AMarotoChivite-cesta-Kotlin/src/main/kotlin/simulacionCesta/factories/CestListFactory.kt
package simulacionCesta.factories

import simulacionCesta.models.CestList
import simulacionCesta.models.Product

object CestListFactory {
    @JvmStatic
    fun create(): CestList {
        // Por defecto presenta un tamaño 10 de máximo
        val randomSize = 10
        val productsStorage = arrayOfNulls<Product>(randomSize)
        do {
            // Introducimos productos aleatorios en la línea de cesta
            for (i in productsStorage.indices) {
                productsStorage[i] = ProductFactory.create()
            }
            // No saldremos hasta que genere por cada línea de producto un producto distinto
        } while (!checkProductNoRepeat(productsStorage))
        return CestList(randomSize, productsStorage)
    }

    /**
     * Verificamos que se generen por cada línea de producto un producto distinto
     *
     * @param productsStorage el almacén que ha generado productos para verificar
     * @return true si no hay productos iguales, false en caso contrario
     */
    fun checkProductNoRepeat(productsStorage: Array<Product?>): Boolean {
        for (i in productsStorage.indices) {
            for (j in i + 1 until productsStorage.size) {
                if (productsStorage[i]!!.name == productsStorage[j]!!.name) {
                    return false
                }
            }
        }
        return true
    }
}